File: algorithms/[690]员工的重要性/solution.js
/**
 * Definition for Employee.
 * function Employee(id, importance, subordinates) {
 *     this.id = id;
 *     this.importance = importance;
 *     this.subordinates = subordinates;
 * }
 */

/**
 * @param {Employee[]} employees
 * @param {number} id
 * @return {number}
 */
const GetImportance = function(employees, id) {
  const map = new Map();
  employees.forEach(({ id, importance, subordinates }) => {
    map.set(id, { value: importance, employee: subordinates });
  });

  let res = 0;
  const queue = [id];
  while (queue.length !== 0) {
    const head = queue.shift();
    const { value, employee } = map.get(head);
    
    res += value;

    if (employee.length) {
      queue.push(...employee);
    }
  }

  return res;
};

// dfs
const GetImportance2 = function(employees, id) {
  const map = new Map();
  employees.forEach(({ id, importance, subordinates }) => {
    map.set(id, { value: importance, employee: subordinates });
  });
  
  let res = 0;
  const dfs = function(k = id) {
    const { value, employee } = map.get(k);
    res += value;
    employee.forEach(e => dfs(e));
  };

  dfs();

  return res;
};

File: algorithms/[690]员工的重要性/solution.py
# topics = ["哈希表", "深度优先搜索"]

from typing import List

"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""


class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates


class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        employees_dict = {e.id: e for e in employees}

        res = 0

        def dfs(k: int) -> None:
            nonlocal res
            employee = employees_dict.get(k)
            if not employee:
                return
            res += employee.importance
            for x in employee.subordinates:
                dfs(x)

        dfs(id)

        return res
